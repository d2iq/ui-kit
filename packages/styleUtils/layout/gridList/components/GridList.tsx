import * as React from "react";
import { css } from "@emotion/css";
import { gridColumnTemplate } from "../style";
import { SpaceSize } from "../../../../shared/styles/styleUtils/modifiers/modifierUtils";
import {
  listReset,
  getResponsiveSpacingStyle
} from "../../../../shared/styles/styleUtils";
import { BreakpointConfig } from "../../../../shared/styles/breakpoints";

interface GridListProps {
  /**
   * How many columns the grid should be. Can be set for all viewport sizes, or configured to have different values at different viewport width breakpoints
   */
  columnCount: BreakpointConfig<number>;
  /**
   * The size of the space between each list item. Can be set for all viewport sizes, or configured to have different values at different viewport width breakpoints
   */
  gutterSize?: SpaceSize;
  /**
   * Which HTML list tag to render the component with
   */
  tag: "ul" | "ol";
  children:
    | Array<React.ReactElement<HTMLLIElement>>
    | React.ReactElement<HTMLLIElement>;
}

const GridList = (props: GridListProps) => {
  const { children, columnCount, gutterSize, tag } = props;
  const GridListEl = tag;

  return (
    <GridListEl
      className={css`
        display: grid;
        ${getResponsiveSpacingStyle("grid-gap", gutterSize)};
        ${gridColumnTemplate(columnCount)};
        ${listReset};
      `}
      data-cy="gridList"
    >
      {children}
    </GridListEl>
  );
};

GridList.defaultProps = {
  gutterSize: "m",
  tag: "ul"
};

export default GridList;
