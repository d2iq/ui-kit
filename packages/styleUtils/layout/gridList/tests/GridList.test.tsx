import React from "react";
import { shallow } from "enzyme";
import serializer from "@emotion/jest";
import toJson from "enzyme-to-json";

import GridList from "../components/GridList";

expect.addSnapshotSerializer(serializer);

describe("GridList", () => {
  it("renders default", () => {
    const component = shallow(
      <GridList columnCount={2} tag="ul">
        <li>list item</li>
        <li>list item</li>
      </GridList>
    );

    expect(toJson(component)).toMatchSnapshot();
  });
  it("renders with all props", () => {
    const component = shallow(
      <GridList columnCount={2} gutterSize="l" tag="ol">
        <li>list item</li>
        <li>list item</li>
      </GridList>
    );

    expect(toJson(component)).toMatchSnapshot();
  });
  it("renders with responsive props", () => {
    const component = shallow(
      <GridList
        columnCount={{ default: 1, medium: 2 }}
        gutterSize={{ default: "s", medium: "m" }}
        tag="ul"
      >
        <li>list item</li>
        <li>list item</li>
      </GridList>
    );

    expect(toJson(component)).toMatchSnapshot();
  });
});
