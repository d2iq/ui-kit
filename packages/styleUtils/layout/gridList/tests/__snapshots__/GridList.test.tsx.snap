// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridList renders default 1`] = `
.emotion-0 {
  display: grid;
  grid-gap: 16px;
  grid-template-columns: repeat(2, minmax(0, 1fr));
  list-style: none;
  margin-left: 0;
  margin-top: 0;
  margin-bottom: 0;
  padding-left: 0;
}

<ul
  className="emotion-0"
  data-cy="gridList"
>
  <li>
    list item
  </li>
  <li>
    list item
  </li>
</ul>
`;

exports[`GridList renders with all props 1`] = `
.emotion-0 {
  display: grid;
  grid-gap: 24px;
  grid-template-columns: repeat(2, minmax(0, 1fr));
  list-style: none;
  margin-left: 0;
  margin-top: 0;
  margin-bottom: 0;
  padding-left: 0;
}

<ol
  className="emotion-0"
  data-cy="gridList"
>
  <li>
    list item
  </li>
  <li>
    list item
  </li>
</ol>
`;

exports[`GridList renders with responsive props 1`] = `
.emotion-0 {
  display: grid;
  list-style: none;
  margin-left: 0;
  margin-top: 0;
  margin-bottom: 0;
  padding-left: 0;
}

@media (min-width: 0) {
  .emotion-0 {
    grid-gap: 12px;
  }
}

@media (min-width: 900px) {
  .emotion-0 {
    grid-gap: 16px;
  }
}

@media (min-width: 0) {
  .emotion-0 {
    grid-template-columns: repeat(1, 1fr);
  }
}

@media (min-width: 900px) {
  .emotion-0 {
    grid-template-columns: repeat(2, 1fr);
  }
}

<ul
  className="emotion-0"
  data-cy="gridList"
>
  <li>
    list item
  </li>
  <li>
    list item
  </li>
</ul>
`;
